### PR TITLE
feat(medusa): integrate pricing module to core

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- integrate pricing module calculate read into pricing service to calculate prices when store endpoints are hit and the pricing module is enabled

### CODE REVIEW

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      return Array.isArray(r.price)
+        ? r.price.map((p) => p.price_set_id)
+        : [r.price.price_set_id]
+    })
+    .flat()
+
+  const queryContext: PriceSelectionContext = removeNullish(context)
+
+  if (queryContext.currency_code) {
+    queryContext.currency_code = queryContext.currency_code.toUpperCase()
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Suggestion : i don't think it should be done jere but maybe the module should always upper case?

File: packages/medusa/src/api/routes/store/products/get-product.ts [hunk 2]
```diff
@@ -134,15 +140,32 @@ export default async (req, res) => {
...
+      const context = await pricingService.collectPricingContext({
         cart_id: validated.cart_id,
         customer_id: customer_id,
         region_id: regionId,
         currency_code: currencyCode,
         include_discount_prices: true,
       })
-    )
+      decoratePromises.push(
+        getProductPricingWithPricingModule(
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/store/products/get-product.ts]: Suggestion : instead of re fetching the variants, should we add a new argument to the product querying function to get the price_set at that moment. The util function would just take care of calculation. Same for the list product

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+      const value = Array.isArray(r.price)
+        ? r.price.map((p) => p.price_set_id)
+        : [r.price.price_set_id]
+      return [r.id, value]
+    })
+  )
+
+  const priceSetIds = res
+    .map((r) => {
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Todo: please can we stop the single letter naming. What is r here since it comes from res, what is res. As a reader you need to go back up to be sure. We agreed last year that single letter was reserved for inline function, and only if the variable before is understandable.

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        price {
+          price_set {
+            id
+          }
+        }
+      }
+    }
+   `
+
+  const res: VariantsRes[] = await remoteQuery(priceSetQuery, {
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Todo: variants_ or variantRes

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    context: variables,
+  })
+
+  const variantIdToPriceSetIdsMap = new Map(
+    res.map((r) => {
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Todo: please no single letter. variant

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  const prices = await pricingModule.calculatePrices(
+    { id: priceSetIds },
+    {
+      context: queryContext,
+    }
+  )
+
+  const priceSetMap = new Map<string, MoneyAmount>(prices.map((p) => [p.id, p]))
+
+  variants.map((v) => {
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Todo: variant

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const priceSetIds = variantIdToPriceSetIdsMap.get(v.id)
+    const prices = priceSetIds?.map((id) => priceSetMap.get(id)).filter(Boolean)
+    if (prices?.length) {
+      const calculatedPrice = prices.reduce((acc: number, curr): number => {
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Todo: curr = price, acc = amount. Naming is so important in general

File: packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        return !curr || acc < curr.amount ? acc : curr.amount
```
Comment (adrien2p) [inline packages/medusa/src/utils/get-product-pricing-with-pricing-module.ts]: Comment: when as user you read this, you have to look all around to know what things are. Harder to debug, fix, read etc. Naming is so important

**Review by adrien2p — CHANGES_REQUESTED** [review]

**Comment by adrien2p** [comment]
Oups sorry i havent seen the draft 😅

File: packages/medusa/src/api/middlewares/error-handler.ts [hunk 9]
```diff
@@ -69,9 +69,9 @@ export default () => {
       case MedusaError.Types.INVALID_ARGUMENT:
         break
       default:
-        errObj.code = "unknown_error"
-        errObj.message = "An unknown error occurred."
-        errObj.type = "unknown_error"
+        // errObj.code = "unknown_error"
+        // errObj.message = "An unknown error occurred."
+        // errObj.type = "unknown_error"
```
Comment (olivermrbl) [inline packages/medusa/src/api/middlewares/error-handler.ts]: todo: Revert

File: packages/link-modules/src/definitions/product-variant-price-set.ts [hunk 10]
```diff
@@ -25,6 +25,7 @@ export const ProductVariantPriceSet: ModuleJoinerConfig = {
       isInternalService: true,
       primaryKey: "id",
       foreignKey: "variant_id",
+      // TODO: Remove this when product module is the default product service
```
Comment (olivermrbl) [inline packages/link-modules/src/definitions/product-variant-price-set.ts]: question: The alias? Is this not related to `isInternalService`?

File: packages/medusa/src/commands/seed.ts [hunk 11]
```diff
@@ -131,6 +134,7 @@ const seed = async function ({ directory, migrate, seedFile }: SeedOptions) {
       categories = [],
       shipping_options,
       users,
+      rule_types,
```
Comment (olivermrbl) [inline packages/medusa/src/commands/seed.ts]: suggestion: Default to `[]`?

File: integration-tests/plugins/helpers/create-default-ruletypes.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import { IPricingModuleService } from "@medusajs/types"
+
+export const createDefaultRuleTypes = async (container) => {
+  const pricingModuleService: IPricingModuleService = container.resolve("pricingModuleService")
+
+  return pricingModuleService.createRuleTypes([{
+    name: 'region_id',
```
Comment (olivermrbl) [inline integration-tests/plugins/helpers/create-default-ruletypes.ts]: todo: needs linting

File: integration-tests/plugins/medusa-config.js [hunk 13]
```diff
@@ -4,7 +4,11 @@ const DB_HOST = process.env.DB_HOST
 const DB_USERNAME = process.env.DB_USERNAME
 const DB_PASSWORD = process.env.DB_PASSWORD
 const DB_NAME = process.env.DB_TEMP_NAME
-const DB_URL = `postgres://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}/${DB_NAME}`
+const DB_URL = `postgres://postgres:@localhost/medusa-integration-1`
+process.env.POSTGRES_URL = DB_URL
+
+const enablePricing = process.env.MEDUSA_FF_PRICING_INTEGRATION == "true"
```
Comment (olivermrbl) [inline integration-tests/plugins/medusa-config.js]:
```suggestion
const enablePricing = process.env.MEDUSA_FF_ISOLATE_PRICING_DOMAIN == "true"
```

File: integration-tests/plugins/medusa-config.js [hunk 14]
```diff
@@ -33,6 +37,8 @@ module.exports = {
     database_extra: { idle_in_transaction_session_timeout: 0 },
   },
   featureFlags: {
+    isolate_product_domain: enableProduct,
+    pricing_integration: enablePricing,
```
Comment (olivermrbl) [inline integration-tests/plugins/medusa-config.js]:
```suggestion
    isolate_pricing_domain: enablePricing,
```

File: packages/medusa/src/loaders/feature-flags/pricing-integration.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { FlagSettings } from "../../types/feature-flags"
```
Comment (olivermrbl) [inline packages/medusa/src/loaders/feature-flags/pricing-integration.ts]: todo(this-file): delete

File: packages/medusa/src/loaders/index.ts [hunk 16]
```diff
@@ -1,43 +1,46 @@
+import "reflect-metadata"
```
Comment (olivermrbl) [inline packages/medusa/src/loaders/index.ts]: todo: rm
```suggestion
import "reflect-metadata"
```

File: packages/medusa/src/loaders/index.ts [hunk 17]
```diff
@@ -167,6 +166,44 @@ export default async ({
     [ContainerRegistrationKeys.MANAGER]: asValue(dataSource.manager),
   })
 
+  container.register("remoteQuery", asValue(null)) // ensure remoteQuery is always registered
+  // Only load non legacy modules, the legacy modules (non migrated yet) are retrieved by the registerModule above
+  if (
+    featureFlagRouter.isFeatureEnabled(IsolateProductDomainFeatureFlag.key) ||
+    featureFlagRouter.isFeatureEnabled(PricingIntegrationFeatureFlag.key)
```
Comment (olivermrbl) [inline packages/medusa/src/loaders/index.ts]: todo: change FF

File: packages/medusa/src/services/pricing.ts [hunk 18]
```diff
@@ -27,6 +39,8 @@ type InjectedDependencies = {
   regionService: RegionService
   priceSelectionStrategy: IPriceSelectionStrategy
   featureFlagRouter: FlagRouter
+  remoteQuery: (...args: any[]) => Promise<any> | null
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: question: Do/should we have a more specific type?

File: packages/medusa/src/services/pricing.ts [hunk 19]
```diff
@@ -160,13 +180,106 @@ class PricingService extends TransactionBaseService {
...
       variantId: string
       quantity?: number
     }[],
     context: PricingContext
   ): Promise<Map<string, ProductVariantPricing>> {
+    if (
+      this.featureFlagRouter.isFeatureEnabled(
+        IsolateProductDomainFeatureFlag.key
+      ) &&
+      this.featureFlagRouter.isFeatureEnabled(PricingIntegrationFeatureFlag.key)
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: change FF

File: packages/workflows/src/handlers/product/update-products-variants-prices.ts [hunk 20]
```diff
@@ -50,10 +54,51 @@ export async function updateProductsVariantsPrices({
...
+        }
+
+        const variantPrices = variantPricesMap.get(variant.id)
+        variantPrices?.push(obj)
+      })
     })
   }
 
-  await productVariantServiceTx.updateVariantPrices(variantIdsPricesData)
+  if (featureFlagRouter.isFeatureEnabled("pricing_integration")) {
```
Comment (olivermrbl) [inline packages/workflows/src/handlers/product/update-products-variants-prices.ts]: todo: change FF

File: packages/medusa/src/utils/remove-nullish.ts [hunk 21]
```diff
@@ -0,0 +1,11 @@
+import { isDefined } from "medusa-core-utils"
+
+export const removeNullish = (
```
Comment (olivermrbl) [inline packages/medusa/src/utils/remove-nullish.ts]: comment: Move to utils?

File: packages/medusa/src/services/pricing.ts [hunk 22]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+      return variant as unknown as PricedVariant
+    })
+  }
+
+  async setAdminProductPricing(
+    products: Product[]
+  ): Promise<(Product | PricedProduct)[]> {
+    if (
+      !this.featureFlagRouter.isFeatureEnabled(
+        PricingIntegrationFeatureFlag.key
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: change FF

File: packages/medusa/src/services/pricing.ts [hunk 23]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+      )
+    ) {
+      return this.setProductPrices(products)
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: Add missing await
```suggestion
      return await this.setProductPrices(products)
```

File: packages/medusa/src/services/pricing.ts [hunk 24]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+  async setAdminVariantPricing(
+    variants: ProductVariant[],
+    context: PriceSelectionContext = {}
+  ): Promise<PricedVariant[]> {
+    if (
+      !this.featureFlagRouter.isFeatureEnabled(
+        PricingIntegrationFeatureFlag.key
+      )
+    ) {
+      return this.setVariantPrices(variants, context)
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: Add missing await
```suggestion
      return await this.setVariantPrices(variants, context)
```

File: packages/medusa/src/services/pricing.ts [hunk 25]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+    return variantIdMoneyAmountMap
+  }
+
+  async setAdminVariantPricing(
+    variants: ProductVariant[],
+    context: PriceSelectionContext = {}
+  ): Promise<PricedVariant[]> {
+    if (
+      !this.featureFlagRouter.isFeatureEnabled(
+        PricingIntegrationFeatureFlag.key
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: Change FF

File: packages/medusa/src/services/pricing.ts [hunk 26]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+        return productVariant as unknown as PricedVariant
+      })
+
+      return product
+    })
+  }
+
+  private async getPricingModuleVariantMoneyAmounts(
+    variantIds: string[]
+  ): Promise<any> {
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]: todo: Add type
```suggestion
  ): Promise<> {
```

File: packages/medusa/src/services/pricing.ts [hunk 27]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+    const query = {
+      product_variant_price_set: {
+        __args: variables,
+        fields: ["variant_id", "price_set_id"],
+      },
+    }
+
+    const variantPriceSets = await this.remoteQuery(query)
+
+    const variantIdToPriceSetIdMap: Map<string, string> = new Map(
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts]:
```suggestion
    const priceSetIdToVariantIdMap: Map<string, string> = new Map(
```

File: packages/medusa/src/services/pricing.ts [hunk 28]
```diff
@@ -509,7 +622,169 @@ class PricingService extends TransactionBaseService {
...
+
+        Object.assign(productVariant, pricing)
+        return productVariant as unknown as PricedVariant
+      })
+
+      return product
+    })
+  }
+
+  private async getPricingModuleVariantMoneyAmounts(
```
Comment (olivermrbl) [inline packages/medusa/src/services/pricing.ts:640]: question: How will this method change as we add more rules? Right now, we've limited possible rule_attributes to `region_id`

**Review by olivermrbl — CHANGES_REQUESTED** [review]

File: packages/medusa/src/services/pricing.ts (hunk 28 above)
Comment (pKorsholm) [reply packages/medusa/src/services/pricing.ts:640]: Good question. At the moment we've prioritized to keep the api unchanged as we don't have support for any other mutations of price sets and rules in the admin either, we've included seeders to create the "default" price rules i.e. `region_id` to allow consistency when migrating to the pricing module. Going forward we need a way of handling additional `rule_attribute`/`value` pairs.

I think a way to handle this in the future could be to extend the `MoneyAmount` response with a `rules` attribute and simply add the pairs there when reading, the values of this object would have to have all possible `ruleType`'s for the `PriceSet` in order to properly validate what can be added and what can't.

CC @riqwan would love your take on this as well

File: packages/medusa/src/services/pricing.ts (hunk 18 above)
Comment (pKorsholm) [reply packages/medusa/src/services/pricing.ts]: added it 😄

File: packages/link-modules/src/definitions/product-variant-price-set.ts (hunk 10 above)
Comment (pKorsholm) [reply packages/link-modules/src/definitions/product-variant-price-set.ts]: yeah, this seems to just be a bad merge

File: packages/medusa/src/services/pricing.ts (hunk 28 above)
Comment (riqwan) [reply packages/medusa/src/services/pricing.ts:640]: Agree with that. We have all the needed information here from this query, we just have to change the shape.

Comment (olivermrbl) [reply packages/medusa/src/services/pricing.ts:640]: Sounds good!

**Review by olivermrbl — APPROVED** [review]
LGTM – great work guys! 🚀

**Comment by olivermrbl** [comment]
Just need a green pipeline :)

**Review by adrien2p — APPROVED** [review]
LGTM 🚀

File: packages/medusa/src/loaders/index.ts [hunk 29]
```diff
@@ -167,6 +164,47 @@ export default async ({
...
+    }
+
+    const medusaApp = await MedusaApp({
+      modulesConfig,
+      servicesConfig: joinerConfig,
+      remoteFetchData: remoteQueryFetchData(container),
+      injectedDependencies,
+    })
+
+    container.register("medusaApp", asValue(medusaApp))
```
Comment (adrien2p) [inline packages/medusa/src/loaders/index.ts]: comment: this will change with phil pr, only the link will be registered in the container instead of everything returned. That way, we don't end up with registrations that we don't really want (we have a lot at the moment) and if medusa app return more things it will not end up in the container un controlled

File: packages/medusa/src/migrations/1697103018534-drop_shipping_profile_profile_id_product_id_unique.ts [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+export const featureFlag = IsolateProductDomain.key
+
+export class dropShippingProfileProfileIdProductIdUnique1697103018534
+  implements MigrationInterface
+{
+  name = "DropShippingProfileProfileIdProductIdUnique1697103018534"
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+      DROP INDEX IF EXISTS "idx_product_shipping_profile_profile_id_product_id_unique";
```
Comment (adrien2p) [inline packages/medusa/src/migrations/1697103018534-drop_shipping_profile_profile_id_product_id_unique.ts]: question: it might sounds dumb but why are we dropping this index?

File: integration-tests/api/medusa-config.js [hunk 31]
```diff
@@ -11,7 +11,7 @@ const enableResponseCompression =
 module.exports = {
   plugins: [],
   projectConfig: {
-    redis_url: redisUrl,
+    // redis_url: redisUrl,
```
Comment (olivermrbl) [inline integration-tests/api/medusa-config.js:14]: question: Was this change intended?
